File: domain.guice/src/main/java/org/openmedstack/domain/guice/ContainerSagaFactory.java
package org.openmedstack.domain.guice;

import com.google.inject.Inject;
import com.google.inject.Injector;
import org.openmedstack.domain.Saga;
import org.openmedstack.eventstore.IConstructSagas;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;

class ContainerSagaFactory implements IConstructSagas {
    private final Injector _container;

    @Inject
    public ContainerSagaFactory(Injector container) {
        _container = container;
    }

    @Override
    public <T extends Saga> T build(Class<T> type, String id) {
        T result = null;
        try {
            var optional = Arrays.stream(type.getConstructors()).findFirst();
            if (optional.isPresent()) {
                var constructor = optional.get();
                var parameters = Arrays.stream(constructor.getParameterTypes()).map(t -> {
                    if (String.class.isAssignableFrom(t)) {
                        return id;
                    }
                    return _container.getInstance(t);
                });
                result = (T) optional.get().newInstance(parameters.toArray());
            }
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
        return result;
    }
}

File: domain.guice/src/main/kotlin/org/openmedstack/domain/guice/ContainerSagaFactory.kt
package org.openmedstack.domain.guice

import com.google.inject.Inject
import com.google.inject.Injector
import org.openmedstack.domain.Saga
import org.openmedstack.eventstore.IConstructSagas
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.util.*

internal class ContainerSagaFactory @Inject constructor(private val _container: Injector) : IConstructSagas {
    override fun <T : Saga> build(type: Class<T>, id: String): T {
        try {
            val constructor: Constructor<T> = type.constructors[0] as Constructor<T>
            val parameters = Arrays.stream(constructor.parameterTypes).map { t: Class<*> ->
                if (String::class.java.isAssignableFrom(t)) {
                    return@map id
                }
                _container.getInstance(t)
            }
            return constructor.newInstance(*parameters.toArray())
        } catch (e: InstantiationException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException(e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException(e)
        }
    }
}
